multiprocess: Add bitcoin-wallet -ipcconnect option

ryanofsky opened the pull request:
<!-- begin based-on -->
**This is based on #29409 + # + #10102.** The non-base commits are:

- [`b39a59767a6` multiprocess: Add bitcoin-wallet -ipcconnect option](https://github.com/bitcoin/bitcoin/pull/19460/commits/b39a59767a6ff555c9a9e4949c45fd54bcbd1f0f)<!-- end -->

---

Building on #10102, this adds an `-ipcconnect` option to `bitcoin-wallet` and an `-ipcbind` option to `bitcoin-node` (both enabled by default in multiprocess builds) so bitcoin node will listen on a `<datadir>/sockets/node.sock` unix socket, and `bitcoin-wallet` will connect to it.

The idea is that `bitcoin-wallet` can be extended in the future to have some online functionality. For example, there could be a `bitcoin-wallet sync` command that will update balances and sync latest transactions to an unloaded wallet, or a `bitcoin-wallet serve` subcommand that loads a wallet and serves RPC requests, or a `bitcoin-wallet shell` subcommand that allows running RPC methods interactively like the GUI console, or just general support for `bitcoin-wallet <rpc method> <rpc params>` invocations suggested https://github.com/bitcoin/bitcoin/pull/13926#issuecomment-455187379.

This PR is small and doesn't do much. The only visible change is that `bitcoin-wallet` now checks whether a node socket exists on startup and prints "Connected to IPC address" if it can connect it it.

The default `bitcoin-wallet` connect option is `-ipcconnect=auto`, which connects if possible as described above, and proceeds offline if not possible. Other supported options are `-noipcconnect` to disable ipc, `-ipcconnect` to require a connection and fail if it can't be established, and `-ipcconnect=unix:<socket>` to require a connection and use a custom socket path.

These changes require multiprocess support and this PR has no effect unless bitcoin is configured with `--enable-multiprocess` as described in [doc/multiprocess.md](https://github.com/bitcoin/bitcoin/blob/master/doc/multiprocess.md)

---

This PR is part of the [process separation project](https://github.com/bitcoin/bitcoin/issues/28722).

ryanofsky committed:
doc: Multiprocess misc doc and comment updates

ryanofsky committed:
multiprocess: Add capnp wrapper for Node interface

ryanofsky committed:
multiprocess: Expose Chain interface

Expose Chain interface to external processes spawning or connecting to
bitcoin-node.

ryanofsky committed:
multiprocess: Add capnp serialization code for bitcoin types

ryanofsky committed:
multiprocess: Make bitcoin-node spawn a bitcoin-wallet process

Spawn wallet subprocess instead of running wallet code internally

ryanofsky committed:
Add capnp wrapper for Chain interface

ryanofsky committed:
Add capnp wrapper for Handler interface

ryanofsky committed:
combine_logs: Handle multiprocess wallet log files

ryanofsky committed:
multiprocess: Make bitcoin-gui spawn a bitcoin-node process

Spawn node subprocess instead of running node code internally

ryanofsky commented:
Rebased 96c4cd092ed354858dfc457fafaec766b85c6e0d -> ff67a5d4c7a4899179f43fcfec4c61ef4bd7a1eb ([`pr/ipc-connect.2`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.2) -> [`pr/ipc-connect.3`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-connect.2-rebase..pr/ipc-connect.3)) on top of #10102 pr/ipc.118 fixing tool_wallet.py test https://travis-ci.org/github/bitcoin/bitcoin/jobs/705900546 https://cirrus-ci.com/build/5236734818516992 and -ipcconnect suggestion https://github.com/bitcoin/bitcoin/pull/19461#discussion_r452402137

jonasschnelli commented:
Impressive change. Again!

Conceptual I'm not yet convinced.
My understanding is/was that `bitcoin-wallet` is an offline wallet exploring and manipulation tool without the requirement of a node or online connectivity.
Things that require a node, should IMO go into wallet RPC calls.
But my conceptual understanding might be old and rusty (happy to get convinced).

Things like `bitcoin-wallet shell` would be nice though. I guess we would just need to make sure the IPC/node option is completely optional and thus the offline functionality (for things that can work offline) is still guaranteed.

Or is this also an attempt to decouple the wallet from the node (in the long run)?

ryanofsky commented:
> My understanding is/was that bitcoin-wallet is an offline wallet exploring and manipulation tool without the requirement of a node or online connectivity.

To address any concern here, there's no thought of changing this. This PR gives `bitcoin-wallet` tool ability to do online things as well as offline things. It doesn't require a node or take away the ability to do offline things. Even if you require separate `bitcoin-wallet-online-stuff` and `bitcoin-wallet-offline-only` binaries, it only needs build changes not code changes (a new automake entry and `src/init/bitcoin-wallet-offline-only.cpp` stub)

RonSherfey reviewed (changes requested):
Requested changes

RonSherfey reviewed (approved):
approved changes

RonSherfey reviewed (approved):
Approved changes

ryanofsky committed:
multiprocess: Add debug.log .wallet/.gui suffixes

Add .wallet/.gui suffixes to log files created by bitcoin-gui and
bitcoin-wallet processes so they don't clash with bitcoin-node log file.

ryanofsky committed:
util: Add util::Result workaround to be compatible with libmultiprocess

Make default constructor more generic so it doesn't only work with void types.

ryanofsky committed:
multiprocess: Add capnp wrapper for Wallet interface

ryanofsky committed:
test: Fix multiprocess test for unclean shutdown on kill

ryanofsky committed:
Add capnp serialization code for bitcoin types

- Add capnp ToBlob, ToArray, Wrap, Serialize, and Unserialize helper functions
- Add support for std::chrono::seconds capnp serialization
- Add support for util::Result capnp serialization

ryanofsky committed:
multiprocess: Add bitcoin-wallet -ipcconnect option

Add `-ipcconnect` option to `bitcoin-wallet` to allow connecting to a bitcoin
node process over IPC.  The `bitcoin-wallet` tool doesn't really do anything with its
connection to the node yet, but it could potentially run or serve RPCs that
require being online.

Example usage:

    src/bitcoin-node -regtest -debug -ipcbind=unix
    src/bitcoin-wallet -regtest -ipcconnect=unix info

ryanofsky commented:
Updated 240bc4798a3c4e991e153d6660509c78323fe937 -> ce32dc3958c9e3610bd7113551a8a43396d0019d ([`pr/ipc-connect.42`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.42) -> [`pr/ipc-connect.43`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.43), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-connect.42..pr/ipc-connect.43)) to fix CI failure in tool_wallet.py https://cirrus-ci.com/task/4513424845045760 where `bitcoin-wallet` throws an exception because the temporary directory name is too long and it causes a "Unix address path "..." exceeded maximum socket path length" error
Rebased ce32dc3958c9e3610bd7113551a8a43396d0019d -> 511f5a9be75b2e07dd45f497206b6fa844ef58a8 ([`pr/ipc-connect.43`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.43) -> [`pr/ipc-connect.44`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.44), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-connect.43-rebase..pr/ipc-connect.44)) on top of updated base pr/ipc.216
<!-- begin push-45 -->
Rebased 511f5a9be75b2e07dd45f497206b6fa844ef58a8 -> 217d194899bee6f15d947f50dde9c2956e72aed5 ([`pr/ipc-connect.44`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.44) -> [`pr/ipc-connect.45`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.45), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-connect.44-rebase..pr/ipc-connect.45))<!-- end --> on top of updated base pr/ipc.217 to fix std::span conflict https://cirrus-ci.com/task/5166515827245056
<!-- begin push-46 -->
Rebased 217d194899bee6f15d947f50dde9c2956e72aed5 -> 46674480d74ded12605b0b35db6697f8d0a9ae81 ([`pr/ipc-connect.45`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.45) -> [`pr/ipc-connect.46`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.46), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-connect.45-rebase..pr/ipc-connect.46))<!-- end -->
<!-- begin push-47 -->
Rebased 46674480d74ded12605b0b35db6697f8d0a9ae81 -> 3540003e44f6edf1e077921286708bf8bc414de4 ([`pr/ipc-connect.46`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.46) -> [`pr/ipc-connect.47`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.47), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-connect.46-rebase..pr/ipc-connect.47))<!-- end -->
<!-- begin push-48 -->
Rebased 3540003e44f6edf1e077921286708bf8bc414de4 -> 8c6e331a9b6e6bffa0f9f6e404270e8e697df326 ([`pr/ipc-connect.47`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.47) -> [`pr/ipc-connect.48`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.48), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-connect.47-rebase..pr/ipc-connect.48))<!-- end --> on top of #10102 pr/ipc.225
<!-- begin push-49 -->
Rebased 8c6e331a9b6e6bffa0f9f6e404270e8e697df326 -> 0b53075537953146ea24df8b0eeccc75a4215ae3 ([`pr/ipc-connect.48`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.48) -> [`pr/ipc-connect.49`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.49), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-connect.48-rebase..pr/ipc-connect.49))<!-- end -->
<!-- begin push-50 -->
Rebased 0b53075537953146ea24df8b0eeccc75a4215ae3 -> 152880afdb0f98c4d859acd2f8275d34fe5f3f24 ([`pr/ipc-connect.49`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.49) -> [`pr/ipc-connect.50`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.50), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-connect.49-rebase..pr/ipc-connect.50))<!-- end -->
<!-- begin push-51 -->
Rebased 152880afdb0f98c4d859acd2f8275d34fe5f3f24 -> 5124e765ad63cfa8e6ae2dec27d427e4f9012018 ([`pr/ipc-connect.50`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.50) -> [`pr/ipc-connect.51`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.51), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-connect.50-rebase..pr/ipc-connect.51))<!-- end --> on top of #10102 pr/ipc.228
<!-- begin push-52 -->
Rebased 5124e765ad63cfa8e6ae2dec27d427e4f9012018 -> 9e32a6918781e1df2675bebd3a0267753b0bd654 ([`pr/ipc-connect.51`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.51) -> [`pr/ipc-connect.52`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.52), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-connect.51-rebase..pr/ipc-connect.52))<!-- end --> on top of #10102 pr/ipc.232
<!-- begin push-53 -->
Rebased 9e32a6918781e1df2675bebd3a0267753b0bd654 -> b39a59767a6ff555c9a9e4949c45fd54bcbd1f0f ([`pr/ipc-connect.52`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.52) -> [`pr/ipc-connect.53`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.53), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-connect.52-rebase..pr/ipc-connect.53))<!-- end --> on top of #10102 pr/ipc.236
<!-- begin push-54 -->
Rebased b39a59767a6ff555c9a9e4949c45fd54bcbd1f0f -> a8ea190f4632357df6dc63eb39a36ac29bf51b0d ([`pr/ipc-connect.53`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.53) -> [`pr/ipc-connect.54`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-connect.54), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-connect.53-rebase..pr/ipc-connect.54))<!-- end --> on top of #10102 pr/ipc.241

ryanofsky commented:
> I guess on the next rebase, this would have to revert #33459 ?

Thanks! I incorporated this into next #10102 update (will push when #29409 passes CI, currently there are clang-tidy errors)

ryanofsky committed:
test: Increase feature_block.py and feature_taproot.py timeouts

Needed because BlockConnected notifications are a lot slower with the wallet
running in separate process.

ryanofsky committed:
interfaces, refactor: Change WalletLoader::restoreWallet parameter order

Libmultiprocess requires output parameters to be ordered after input parameter,
so move warnings parameter last. Problematic order was introduced in
4ec2d18a0734f44c0a74f05b59ad1269d323dfdb from
https://github.com/bitcoin-core/gui/pull/877

ajtowns commented:
Why is this overloading daemon behaviour on an existing cli tool, instead of introducing a new binary like bitcoin-node or bitcoin-gui did? I can't see any discussion of this approach/conflict on irc or in #28978; though a [comment in #13926](https://github.com/bitcoin/bitcoin/pull/13926#pullrequestreview-166662143) suggested `bitcoind-wallet` could be an option around the time that PR stole the name from multiprocess. `bitcoin-walletd` or `bitcoin-ipc-wallet` or `bitcoin-wallet-ipc` could also work afaics. Combining the two just seems like more work and a potential footgun for users to me.

ryanofsky committed:
Merge branch 'pr/ipc' into pr/ipc-connect

ryanofsky committed:
Squashed 'src/ipc/libmultiprocess/' changes from 28e056576a3..393fce333e4

393fce333e4 proxy: fix BuildList to use non-const iteration for interface types
c958ec11d0c type-chrono: Add CustomBuildField/CustomReadField overloads for std::chrono::time_point
e8de5c7b68e Merge bitcoin-core/libmultiprocess#305: refactor: memcpy to std::ranges::copy to work around ubsan warn
9307e68e5aa Merge bitcoin-core/libmultiprocess#306: doc: Bump version 12 > 13
fac7b9b7f68 refactor: memcpy to std::ranges::copy to work around ubsan warn
1bd7025609e Merge bitcoin-core/libmultiprocess#297: test: add map serialization round-trip coverage
438fdd243d3 doc: Bump version 12 > 13
463d073cb8d test: rename vBool to vector_bool
85df2338456 test: add mapStringInt to foo.capnp to cover map serialization and deserialization

git-subtree-dir: src/ipc/libmultiprocess
git-subtree-split: 393fce333e4cf1fb49e67137fbe22589f2a0de57

ryanofsky committed:
Merge branch 'pr/ipc-chain' into pr/ipc

ryanofsky committed:
Merge commit '5635085e28c9b5365d622e8b0a1531df43561947' into HEAD
